devolutions-gateway/src/http/controllers/kdc_proxy.rs

```diff
@@ -0,0 +1,160 @@
+use picky_krb::messages::KdcProxyMessage;
+use saphir::controller::Controller;
+use saphir::http::Method;
+use saphir::macros::controller;
+use saphir::request::Request;
+use saphir::response::Builder;
+use std::net::{SocketAddr, ToSocketAddrs};
+use std::sync::Arc;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::{TcpStream, UdpSocket};
+
+const ERROR_BAD_FORMAT: &str = "\x0b";
+const DEFAULT_KDC_PORT: u16 = 88;
+
+pub struct KdcProxyController {
+    config: Arc<Config>,
+}
+
+impl KdcProxyController {
+    pub fn new(config: Arc<Config>) -> Self {
+        Self { config }
+    }
+}
+
+#[controller(name = "KdcProxy")]
+impl KdcProxyController {
+    #[post("/")]
+    async fn proxy_kdc_message(&self, req: Request) -> Result<Builder, HttpErrorStatus> {
+        let kdc_proxy_message = KdcProxyMessage::from_raw(
+            req.load_body()
+                .await
+                .map_err(|_| HttpErrorStatus::bad_request(ERROR_BAD_FORMAT))?
+                .body(),
+        )
+        .map_err(|_| HttpErrorStatus::bad_request(ERROR_BAD_FORMAT))?;
+
+        let realm = if let Some(realm) = &kdc_proxy_message.target_domain.0 {
+            realm.0.to_string()
+        } else {
+            return Err(HttpErrorStatus::bad_request(ERROR_BAD_FORMAT));
+        };
+
+        let kdc_proxy_config = self
+            .config
+            .kdc_proxy_config
+            .as_ref()
+            .ok_or_else(|| HttpErrorStatus::internal("KDC proxy is not configured"))?;
+
+        if kdc_proxy_config.reaml != realm {
+            return Err(HttpErrorStatus::bad_request("Requested domain is not supported"));
+        }
+
+        let scheme = kdc_proxy_config.kdc.scheme();
+        let address_to_resolve = format!(
+            "{}:{}",
+            kdc_proxy_config.kdc.host().unwrap().to_string(),
+            kdc_proxy_config.kdc.port().unwrap_or(DEFAULT_KDC_PORT)
+        );
+
+        let kdc_address = if let Some(address) = lookup_kdc(&address_to_resolve) {
+            address
+        } else {
+            slog_scope::error!("Unable to locate KDC server");
+            return Err(HttpErrorStatus::internal("Unable to locate KDC server"));
+        };
+
+        let mut kdc_reply_message = Vec::new();
+
+        if scheme == "tcp" {
+            let mut connection = TcpStream::connect(kdc_address).await.map_err(|e| {
+                slog_scope::error!("{:?}", e);
+                HttpErrorStatus::internal("Unable to connect to KDC server")
+            })?;
+
+            connection
+                .write_all(&kdc_proxy_message.kerb_message.0 .0)
+                .await
+                .map_err(|e| {
+                    slog_scope::error!("{:?}", e);
+                    HttpErrorStatus::internal("Unable to send the message to the KDC server")
+                })?;
+
+            read_kdc_reply_message(&mut connection, &mut kdc_reply_message).await?;
+        } else if scheme == "udp" {
+            let mut buff = vec![0; 1024];
```

Comment:
Can you explain why this buffer has this size? Is 1024 the upper bound for a response in UDP mode?
Also, you may use an array instead of a vec: `[0; 1024]`

Reply:
I thought that a ticket could not be more than 1K. I rechecked and the upper bound for a response in UDP mode is 64K. 
I think 2k of a reziable buffer would be enough.

Reply:
Yeah 64K sounds like a lot. You can bump to 2K or stay at 1K, but add a comment about that maybe :+1: